composer/distributed/fsdp2.py

```diff
@@ -168,79 +279,64 @@ def update_optimizer_modules(
 
     Returns:
         None
     """
+    # Define the default kwargs for fully_shard
     fully_shard_kwargs = {'mesh': fsdp2_config.device_mesh, 'reshard_after_forward': fsdp2_config.reshard_after_forward}
 
-    # Apply fully_shard to each module in the list
-    if len(independent_submodules) == 0:
-        raise RuntimeError(
-            "Can't find any submodules to apply FSDP, e.g., the submodules may all have tied parameters. Applying FSDP to the root model does not provide any memory savings.",
-        )
+    # Apply fully_shard to each relevant module defined by the policy
+    _recursive_apply_fully_shard(model, model, fsdp2_config, auto_wrap_policy, fully_shard_kwargs)
 
-    independent_submodules, modules_tied = get_standalone_and_tied_modules(independent_submodules)
-    if len(modules_tied) > 0:
-        raise RuntimeError(
-            'Submodules to be sharded have tied parameters. FSDP cannot be applied to modules with tied parameters independently. '
-            'Please ensure that the submodules do not have tied parameters.',
-        )
-
-    # NOTE there is a bug fully_shard can not handle when the model has a child module which is the child of another
-    # to be FSDPed child module. For example:
-    # model
-    #  self.child1
-    #  self.child2
-    #    ├── self.child1
-    #    └── grandchild
-    # We can fully_shard self.child2 however if we call fully_shard on `model` again, then
-    # due to `child1` is not a FSDPModule, it will try to fully_shard params in `child1` which is already sharded by `child2`,
-    # and it errors out with misleading error as it thinks it is applying FSDP on top of another parallelism.
-
-    # Currently identify_shardable_modules avoids this case through generally handling weight tying so that only parent of child1 and child2
-    # is sharded. However if we allow users to call this function directly with custom modules_to_shard, we need to:
-    # legalize that no module outside modules_to_shard shares parameters with modules_to_shard or
-    # TODO alternatively we can fix torch/distributed/fsdp/_fully_shard/_fsdp_init.py::_get_managed_modules
-    legalize_param_sharing_between_modules(model, independent_submodules)
+    # Get the wrapping policy of the root model and apply fully_shard as specified
+    root_policy_decision = auto_wrap_policy(model)
 
-    fully_shard(independent_submodules, **fully_shard_kwargs)
-    # Apply fully_shard to the parent model to ensure all parameters are sharded
-    fully_shard(model, **fully_shard_kwargs)
+    # We default to wrapping the root module as long as its wrapping policy is not explicitly set to False
+    if root_policy_decision is True or root_policy_decision is None:
+        fully_shard(model, **fully_shard_kwargs)
+    elif isinstance(root_policy_decision, dict):
+        kwargs = root_policy_decision
+        fully_shard(model, **kwargs)
+    else:
```

Comment:
should the else be an error? if not, can just delete, don't need an else that just has a pass

Reply:
Updated this to follow the same mechanics as `FSDP1` and added a comment about `dist_strategy.py:L367`